spark-job-server-proposal.md

```diff
@@ -0,0 +1,75 @@
+# Adding Spark Job Server To Run Spark Jobs
+
+## Introduction
+
+This document provides design of spark job server which would trigger spark
+jobs either as scheduled or batch jobs. Providing the user a ui for inputing
+the source code git location, environment variables required to integrate with
+other apps, input parameters for spark job, selecting s2i builder image and
+files required.
+
+## Problem Statement
+
+The Data Platform Spark applications can only run within OpenShift
+environment and users are required to do provide only source code which
+currently we use source to image to create a standard image which will be run
+against the spark cluster. Currently there is no way to schedule jobs or to run
+cron jobs. Also if users have a more complex setup where they would need
+JDG or other servers they would be required to do a lot of manual setup.
+
+## Proposed Solution
+
+Apache Spark users will submit their spark job in the form of binary jar or py
+files. These files will require users to provide input as well as be triggered
+by spark-submit shell script. However there are other environment requirements
+that may be needed by users as well.
+
+The requirements to create, run, rerun, stop a particular spark job by one
+click of a button.
+
+With a simple html form users can provide the following input:
+* source git url
+* input parameters for spark job
+* files required to be mounted to run the job.
+* inject environment variables from services that this spark job would need
+to integrate with.
+* specify if they would like a new cluster created or if they would like to use
+an existing one
+* schedule when the job should run
+* how often the job should run
+
+The workflow would be user fills out html form and oshinko-console would send
+json payload with the information required to build the s2i image and then run
+a spark job including the inputs provided above. After that we would get an
+update in the ui to see jobs running and their status. There would be a button
+on the ui for users to rerun the spark job.
+
+
+## Alternatives
+
+Alternative is to utilize templates to trigger a build which limits users to
+only be able to create a single container with binary that runs the spark job.
+
+
+## Affected Components
+
+- oshinko-rest : Will need to have server-side rest api for
+create/list/start/restart/delete Spark job within kubernetes and have error
+handling incase of failure scenario. Also the server-side code would need to
+integrate with kubeapi. The backend when creating a spark job would first need
+to create the image using s2i technology then run the image in a batch/scheduled
+job against the kube api.
+- oshinko-console : html form, angularjs javascript to make rest calls to
+backend api that would create/list/start/restart/delete spark jobs.
+
```

Comment:
i'd like more discussion about the scope of work here. as it stands, the spec asks for an html form that will be created and all interaction is through that piece. but here we are talking about parity between the console project and rest interface. these changes would only exist in the rest interface without a clear way to interact with them(will that be through the forthcoming cli?).

i'd like to get @crobby and @sub-mod to comment on the scope of work for the -console portion of this.

my concern here is that we currently have 2 different UI applications, the oshinko-console and oshinko-webui. webui is sticking around as a development tool for the moment, but will we need to have the job interface in the webui or will it be console only?